### PR TITLE
React Pages Migration: Conditions

### LINKED ISSUE #3570: Improve UI for Assessment Settings (unlock conditions)
On Edit Assessment page, user is frustrated when earlier entered assessment settings disappear or reverts because he sets Unlock Conditions. Perhaps the assessment setting should be preserved or saved before bring the user to the Unlock Conditions page.

### LINKED ISSUE #4242: Move views for creating and updating conditions from RoR to React
Screenshot 2021-12-27 at 12 56 55 PM

Currently, when editing a conditional, the conditions list appears as a React component.

Screenshot 2021-12-27 at 12 59 56 PM

However, when performing actions on conditions themselves (e.g. creating or updating a condition), the app navigates to a separate page where an RoR view for the relevant action is rendered.

We want to move the RoR views into React so that they can be reused (currently, each condition type has its own set of RoR views) and also remove page navigation (i.e. creating and editing conditions will take place on the conditional editing page).

### PR DESCRIPTION
**Pull request by purfectliterature**
Part of #5098. Closes #3570. Closes #4242.

| Before | After |
| - | - |
| image | image |
| image | image |
| image | image |
| image | image |

## Notable changes
- 🔥 New `useToggle` hook for (annoying) boolean `useState`s.  
`onClick={(): void => setLoading(false)}` becomes `onClick={toggleLoading}`!
- 🔥 New `Preload` component that abstracts the `useEffect`/`isLoading` combo that allows a declarative way of defining loading and main components, and giving the main components _guaranteed non-`undefined` fetched data_.
- 🔨 Add `@tailwindcss/line-clamp` plugin for Tailwind for [multi-line truncations](https://tailwindcss.com/blog/multi-line-truncation-with-tailwindcss-line-clamp).
- ~~`AssessmentForm` now loads form data via JSON instead of the 🤮 `data=` attribute.~~ Moved to #5115.

## ✨ Cool things ✨
- Achievement condition's `Autocomplete` shows the badges and descriptions of all course achievements. Users can also search by either the achievement title or description.
- Literally this 👇

## On `Preload` for data fetching
I grew sick of seeing the `useEffect`/`isLoading` combo for data fetching, especially with `Suspense` out of the question as of #5115. Handling the `undefined` data in my `*Condition.tsx` was difficult to do in one component, because 
- `useForm` and hooks cannot be put under if-clauses, but 
- it needs the non-`undefined` fetched data as part of its `defaultValues`. `reset`-ing after the `useEffect` finishes just makes the code unreadable, and still does not solve the typechecking issue with `undefined`-able data.

Enter `Preload`. It is a declarative way of fetching data and rendering loading and main components. Given that `fetchData: () => Promise<Data>`, instead of doing this ❌❌❌❌ 👎👎👎👎 :
```tsx
const [data, setData] = useState<Data>();
const [loading, setLoading] = useState(true);

useEffect(() => {
  fetchData().then((fetchedData) => {
    setData(fetchedData);
    setLoading(false);
  };
}, []);

if (loading) return <LoadingIndicator />;

return <MainComponent data={data} />;
```
>Note that we use `loading` instead of just checking `!data` because if `data` is of type `Record` (POJO) and has an initial value of `{}` (so that `data` type will not be `Data | undefined`), `!data` is actually `false` because `Boolean({})` is `true`. [Checking for emptiness of an object is not as trivial as one might think](https://stackoverflow.com/questions/21743921/javascript-is-an-empty-object-a-falsy-one).

we can simply do this with `Preload` ✅✅✅✅ 👍👍👍👍 :
```jsx
return (
  <Preload while={fetchData} render={<LoadingIndicator />}>
    {(data): JSX.Element => (
      <MainComponent data={data} />
    )}
  </Preload>
)
```
Bonus! `data` that is given to `MainComponent` is guaranteed to be non-`undefined` and its type is inferred automatically from `fetchData`'s return type. How cool is that? 😎 

Some props you might use other than `while` and `render`:
- `silently: boolean` allows you to not have any toasts if there is any fetching errors.
- `onErrorToast: string` allows you to define your own toast message if there is any fetching errors.
- `onErrorDo: (error) => void` allows you to grab the error response data and do anything you want with it.

This syntax is inspired by `Suspense`, and is defined in a (relatively) imperative syntax: _while fetchData, render LoadingIndicator, then when data is ready, render the MainComponent_. Forget `useEffect` 🤮 and that `loading` or `isLoading` state 🤧 polluting your code and triggering Code Climate all the time. Use `Preload` 🤟❤️.

On the surface, `Preload` might just seem like a simple abstraction, and not really an innovation. Wrong 🙅. `Preload` here paves our future with `Suspense`. I personally believe that when `Suspense` for data fetching is stable for production and its libraries, e.g., react-query, SWR, etc. are good, we should move to it for a more efficient and responsive front-end. Using `Preload` now means when that bright future 🌤️ comes, we can just do a one-time-good-one move from `useEffect`/`isLoading` to `Suspense` by modifying `Preload`.

One can also add ⏲️ debouncing, ⬇️ auto-revalidations, 📦 caching, or connections to ♻️ Redux store in `Preload`, and all fetchers in Coursemology will benefit from it.

Remember, it essentially abstracts data fetching out from our view components! Your component can now focus on being absolutely fricken' cool 😎 without having to deal with `undefined` and `loading` states.

### CODE REVIEW

**Comment by purfectliterature** [comment]
Remaining specs to resolve:
`spec/features/course/assessment_condition_management_spec.rb`
`spec/features/course/achievement_condition_management_spec.rb`

**Review by ekowidianto — CHANGES_REQUESTED** [review]
Great work as always! Some comments below:
- [x] Is _assessment_condition.html.slim still used anywhere else?
- [x] It feels a bit misleading to put the `ConditionsManager` folder inside `form/fields`. Do you think it would be better to put it under `lib/coponents/course/condition` or somewhere else? What do you think?
- [ ] When retrieving conditions (achievement, assessment and survey), you should only retrieve the available ones (eg see achievements/_form/html.slim line 3, and the same for assessment and survey). This will prevent the obvious illegal condition to be set in the first place. Also, instead of re-using the index APIs for achievement, assessment and survey, you can create new API endpoints to retrieve these conditions (via each specific condition controller - we can call it `def available_achievements` for example). This way, we dont need to retrieve all unnecessary information via the index endpoint which will slow down the query process since we only need the `id` and `title` anyway. (bonus: you dont need to filter out its own condition in the frontend anymore).
- [ ] Sort the list of achievement, assessment and survey by the name.
- [ ] Using 1010s data, I get some key error warnings for achievement condition.
- [ ] When there is an error after a condition deletion, there is no error message and it fails silently. 
- [ ] For create/update, if the error returned is not of validation error (eg network error, 403/404, error etc), add a toast error message. Otherwise, it would fail silently. Also `errors.errors` may be undefined, as such we need to check for it before calling `setError`

File: client/app/lib/components/form/fields/ConditionsManager/conditions/AnyCondition.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { ConditionData } from 'types/course/conditions';
```
[the remark this answers is not in the document]
Comment (purfectliterature) [reply client/app/lib/components/form/fields/ConditionsManager/conditions/AnyCondition.ts:1]: I only put types that represent our app's business logic, e.g., models, backend conformers, etc. In this case, `AnyCondition` is a 'generic' interface that `*Condition.tsx`s in the future should use, i.e., it is just a component's (props) supporting interface.

Declaring these types in `conditions/index.tsx` will cause a dependency cycle:
- `*Condition.tsx`s import `AnyCondition` from `conditions/index.tsx`, and
- `specifiers.ts` imports `*Condition.tsx` and `AnyCondition` from `conditions/index.tsx`,

hence resulting in dependency cycles in `*Condition.tsx`s via `specifiers.ts`.

Due to these reasons, I made a separate `AnyCondition.ts` in `conditions`.

**Comment by purfectliterature** [comment]
To continue from #5115:
- [ ] Consolidate `Prompt` and `ConditionDialog`
- [ ] Deprecate `ConfirmationDialog`
- [ ] Convert all `useEffect`/`isLoading` to `Preload` for instant fetch error catching
- [ ] Implement global `Form` `disable` for UI freezing
- [ ] `Form` heads-up to freeze or use `LoadingButton`
- [ ] Convert all boolean states to `useToggle`

File: client/app/lib/components/form/fields/ConditionsManager/index.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+  return (
+    <ConditionsProvider value={conditionContext}>
+      <Subsection
+        title={props.title}
+        subtitle={props.description}
+        className="mt-4"
+      >
+        <div className="flex h-16 items-center space-x-4">
+          <Button
```
[the remark this answers is not in the document]
Comment (purfectliterature) [reply client/app/lib/components/form/fields/ConditionsManager/index.tsx]: Sure, but I was not really sure why we need to disable it. 

When the menu is open, it is in a separate presentation overlay, so clicking anywhere outside the menu will first dismiss the menu, hence double-clicking the button is impossible. And if for whatever reason it becomes possible, `toggleAdding` will just set `adding` to `false` and the menu will disappear (as expected).

File: client/app/lib/components/form/fields/ConditionsManager/conditions/ConditionDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+          disabled={props.primaryActionDisabled}
+        >
+          {props.primaryAction}
+        </Button>
+      </DialogActions>
+    </Dialog>
+  );
+};
+
+export default ConditionDialog;
```
[the remark this answers is not in the document]
Comment (purfectliterature) [reply client/app/lib/components/form/fields/ConditionsManager/conditions/ConditionDialog.tsx]: `ConditionDialog` is now removed. `*Condition` now uses `Prompt`, and `Prompt` itself now supports more advanced customisations.

**Review by ekowidianto — APPROVED** [review]
Thanks for the changes!

**Pull request merged**